### PR TITLE
feature(api): add authentication for api routes

### PR DESCRIPTION
**Pull request by jmanuelrosa**
Add an authentication system based on Google Oauth.

To access the dashboard (aka app) to book a shift, register food, etc, a volunteer (or a user) must do login or, in other words, be authenticated. Added authentication and authorization steps in our API to keep this access private for every user with an account that they can use to be authenticated and authorized.

### How it works

A user can do login in to our application using Google as a provider. Right now is the only way to do it, so in this case, we keep the same flow that we have nowadays.

When a user accesses, we delegate this process in Google Oauth. If it's ok, we create an internal token (based on jwt) and save it in a cookie. 

With a cookie, every time that we make a request to the API, the browser will add the cookie as part of it, so in the API the process to check the authorization is easy. In the frontend, we reduce the code a lot because the way to send the token to the API is automatic by the browser

### Considerations
As the project is based on the web platform, using a cookie to keep the session is the easiest and most secure way to keep the information about the authentication. Cookie set in the server, based on HTTPS, secure, signed and based on lax SameSite rules (to be able to send from the app website to the API). In this case, nobody can edit the cookie (as they can do with the localStorage) to edit or try to attack our API.

If a user can't access because of authorization or authentication, we send a 401 error based on authentication or 503 based on service unavailable in case Google fails.

**Note**: We will need to revisit this decision or provide a new way for the authentication in case we would like mobile devices to have access to the API.

### CODE REVIEW

**Comment by jmanuelrosa** [comment]
@madboks/devs I just updated and changed the logic to authenticate in the app instead of on the web. I added some routes as an example while we don't work on the app interface.

File: apps/web/src/pages/index.astro [hunk 1]
```diff
@@ -1,7 +1,10 @@
 ---
 import Layout from '../layouts/default.astro';
 ---
-
 <Layout>
   Public website for Madboks
+
+  <p>
+    <a href={import.meta.env.PUBLIC_APP_URL}>Sign up/Sign in</a>
+  </p>
```
Comment (gisemartinez) [inline apps/web/src/pages/index.astro:9]: praise: Yay! login button ready!

File: apps/api/src/services/auth/google.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    error: string
+    code: string
+    redirect: string
+  }
+}>
+
+async function fetchUserInfo (code: string): Promise<Response> {
+  // TODO: Add pkce login check
+  try {
+    const response = await fetch('https://oauth2.googleapis.com/token', {
```
[the remark this answers is not in the document]
Comment (jmanuelrosa) [reply apps/api/src/services/auth/google.ts]: For a test environment we will mock the call using an spy like Sinon or another technic offered from the library that we would like to use, like [mockImplementationOnce](https://jestjs.io/docs/mock-function-api#mockfnmockimplementationoncefn), for example. Another option is to set this URL as an env var, but the problem with this is that I don't thinkt hat it's a good idea to finish with an env vars hell, when we don't reuse this url in any place of the app, for example.

File: apps/api/src/lib/auth.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      throw new AuthError({ cause: 'INVALID_TOKEN' })
+    }
+
+    return {
+      id: payload.sub as JWTVerify['sub'],
+      email: payload.email as JWTVerify['email'],
+    }
+  } catch (error) {
+    throw new AuthError({ cause: 'INVALID_TOKEN' })
+  }
```
[the remark this answers is not in the document]
Comment (jmanuelrosa) [reply apps/api/src/lib/auth.ts]: I think that it's not a good idea to expose the real error. It's good to catch them and expose a generic error, but in terms of authentication user's doesn't need to know that the jwt token is expired or cookie is not valid. So, for security reasons we can expose a generic error that for us, trigger the same functionality, for example:

Foe us, INVALID_TOKEN means that the JWT or the cookie is expired. As the cookie is the place that we storage the jwt token, for us, this error means that user is not able to be authorized so we need to regenerate the token (and user need to do login, or whatever, for example)

File: apps/api/src/plugins/authorization.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  if (!session) {
+    throw new AuthError({ cause: 'INVALID_TOKEN' })
+  }
+
+  const cookie = req.unsignCookie(session)
+
+  if (!cookie.valid || !cookie.value) {
+    throw new AuthError({ cause: 'INVALID_TOKEN' })
+  }
+
```
[the remark this answers is not in the document]
Comment (jmanuelrosa) [reply apps/api/src/plugins/authorization.ts]: I just add a constant directory when we can save all this information

**Review by gisemartinez — APPROVED** [review]

**Pull request merged**